File: prediction-market-builder/backend/app/models/strategy.py
import uuid
import enum
from datetime import datetime, timezone
from sqlalchemy import String, DateTime, Float, JSON, Text, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column
from app.database import Base


class StrategyStatus(str, enum.Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    ARCHIVED = "archived"


class Strategy(Base):
    __tablename__ = "strategies"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(String, nullable=False, index=True)
    name: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=True)
    status: Mapped[StrategyStatus] = mapped_column(SAEnum(StrategyStatus), default=StrategyStatus.DRAFT)
    mode: Mapped[str] = mapped_column(String, default="chat")
    nodes: Mapped[list] = mapped_column(JSON, default=list)
    edges: Mapped[list] = mapped_column(JSON, default=list)
    risk_profile: Mapped[dict] = mapped_column(JSON, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

File: frontend/src/types/strategy.ts
export type StrategyStatus = 'draft' | 'active' | 'paused' | 'archived'
export type StrategyMode = 'chat' | 'node' | 'hybrid' | 'automated'

export interface StrategyNode {
  id: string
  type: string
  position: { x: number; y: number }
  data: Record<string, unknown>
}

export interface StrategyEdge {
  id: string
  source: string
  target: string
}

export interface Strategy {
  id: string
  name: string
  description?: string
  status: StrategyStatus
  mode: StrategyMode
  nodes: StrategyNode[]
  edges: StrategyEdge[]
  risk_profile: Record<string, unknown>
  created_at: string
  updated_at: string
}
